File: backend/app/services/analyzer.py
"""分析引擎 - 调用大模型 API 进行资产分析"""

import os
from openai import OpenAI
from typing import Optional


SKILLS_DIR = os.path.join(os.path.dirname(os.path.dirname(__file__)), "skills")


def _load_skill(skill_name: str) -> str:
    """加载 skill 文件内容"""
    path = os.path.join(SKILLS_DIR, f"{skill_name}.md")
    with open(path, "r", encoding="utf-8") as f:
        return f.read()


def _get_client() -> OpenAI:
    """获取 OpenAI 客户端，支持自定义 base_url"""
    api_key = os.getenv("OPENAI_API_KEY", "")
    base_url = os.getenv("OPENAI_BASE_URL", None)
    return OpenAI(api_key=api_key, base_url=base_url)


def _call_llm(system_prompt: str, user_prompt: str) -> str:
    """调用大模型 API"""
    client = _get_client()
    model = os.getenv("OPENAI_MODEL", "gpt-4o")
    response = client.chat.completions.create(
        model=model,
        messages=[
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": user_prompt},
        ],
        temperature=0.7,
        max_tokens=4000,
    )
    return response.choices[0].message.content


def analyze_assets(holdings_text: str, feedback: Optional[str] = None) -> str:
    """资产配置分析"""
    skill = _load_skill("asset_analysis")
    user_prompt = f"请分析以下客户持仓数据：\n\n{holdings_text}"
    if feedback:
        user_prompt += f"\n\n客户经理反馈意见（请据此调整分析）：\n{feedback}"
    return _call_llm(skill, user_prompt)


def analyze_trades(trades_text: str, feedback: Optional[str] = None) -> str:
    """交易行为分析"""
    skill = _load_skill("trade_behavior")
    user_prompt = f"请分析以下客户交易记录：\n\n{trades_text}"
    if feedback:
        user_prompt += f"\n\n客户经理反馈意见（请据此调整分析）：\n{feedback}"
    return _call_llm(skill, user_prompt)


def generate_summary(
    asset_analysis: str,
    trade_analysis: str,
    customer_name: str = "客户",
    feedback: Optional[str] = None,
) -> str:
    """生成综合报告"""
    skill = _load_skill("report_template")
    user_prompt = (
        f"客户名称：{customer_name}\n\n"
        f"## 资产配置分析结果\n\n{asset_analysis}\n\n"
        f"## 交易行为分析结果\n\n{trade_analysis}"
    )
    if feedback:
        user_prompt += f"\n\n客户经理反馈意见（请据此调整报告）：\n{feedback}"
    return _call_llm(skill, user_prompt)


def run_full_analysis(
    holdings_text: str,
    trades_text: str,
    customer_name: str = "客户",
    feedback: Optional[str] = None,
) -> dict:
    """运行完整分析流程"""
    asset_result = analyze_assets(holdings_text, feedback)
    trade_result = analyze_trades(trades_text, feedback)
    summary = generate_summary(asset_result, trade_result, customer_name, feedback)

    return {
        "asset_analysis": asset_result,
        "trade_analysis": trade_result,
        "summary": summary,
    }

File: backend/services/analyzer.js
/**
 * 分析引擎 - 调用 Anthropic 兼容 API 进行资产分析
 */
const Anthropic = require("@anthropic-ai/sdk");
const fs = require("fs");
const path = require("path");

const SKILLS_DIR = path.join(__dirname, "..", "app", "skills");

const API_KEY = "sk-cp-Ky8-iKvBLG2Ot8YEIEKhn0_OeqEgedkr7JljIl36E5NIQeM8sA6Bku3gDBjNrLa4XimgzQHFoSnTdDNwYIlPZcYcVFx_tK4NA-Xp9WAQF94i5ihJORjDebg";
const BASE_URL = "https://api.minimaxi.com/anthropic";
const MODEL = "MiniMax-M2.7";

function loadSkill(skillName) {
  const filePath = path.join(SKILLS_DIR, `${skillName}.md`);
  return fs.readFileSync(filePath, "utf-8");
}

function getClient() {
  return new Anthropic({
    apiKey: API_KEY,
    baseURL: BASE_URL,
  });
}

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function callLLM(systemPrompt, userPrompt, maxRetries = 10) {
  const client = getClient();

  for (let attempt = 1; attempt <= maxRetries; attempt++) {
    try {
      const response = await client.messages.create({
        model: MODEL,
        max_tokens: 4000,
        system: systemPrompt,
        messages: [
          { role: "user", content: userPrompt },
        ],
      });
      console.log("API 响应结构:", JSON.stringify({
        stopReason: response.stop_reason,
        hasContent: !!response.content,
        contentLength: response.content?.length,
        contentTypes: response.content?.map(c => c.type),
      }));
      
      // MiniMax 返回的是 thinking 类型，需要查找 text 类型的内容
      const textContent = response.content?.find(c => c.type === "text");
      const result = textContent?.text;
      
      if (!result || result.trim().length === 0) {
        console.log(`API 返回空内容，内容类型: ${response.content?.map(c => c.type).join(", ")}`);
        throw new Error("API 返回空内容");
      }
      return result;
    } catch (err) {
      const status = err.status || err.statusCode || (err.error && err.error.status);
      const isRetryable = status === 529 || status === 503 || status === 429 || err.message === "API 返回空内容";
      if (isRetryable && attempt < maxRetries) {
        // 基础间隔5秒，加随机抖动，最长60秒
        const baseDelay = 5000 * Math.pow(1.5, attempt - 1);
        const jitter = Math.random() * 3000;
        const delay = Math.min(baseDelay + jitter, 60000);
        console.log(`第${attempt}/${maxRetries}次重试，${(delay / 1000).toFixed(1)}秒后重试... 原因: ${err.message || status}`);
        await sleep(delay);
      } else {
        throw err;
      }
    }
  }
  throw new Error("大模型 API 调用失败，已达最大重试次数");
}

async function analyzeAssets(holdingsText, feedback) {
  const skill = loadSkill("asset_analysis");
  let userPrompt = `请分析以下客户持仓数据：\n\n${holdingsText}`;
  if (feedback) {
    userPrompt += `\n\n客户经理反馈意见（请据此调整分析）：\n${feedback}`;
  }
  return callLLM(skill, userPrompt);
}

async function analyzeTrades(tradesText, feedback) {
  const skill = loadSkill("trade_behavior");
  let userPrompt = `请分析以下客户交易记录：\n\n${tradesText}`;
  if (feedback) {
    userPrompt += `\n\n客户经理反馈意见（请据此调整分析）：\n${feedback}`;
  }
  return callLLM(skill, userPrompt);
}

async function generateSummary(assetAnalysis, tradeAnalysis, customerName, feedback) {
  const skill = loadSkill("report_template");
  let userPrompt =
    `客户名称：${customerName}\n\n` +
    `## 资产配置分析结果\n\n${assetAnalysis}\n\n` +
    `## 交易行为分析结果\n\n${tradeAnalysis}`;
  if (feedback) {
    userPrompt += `\n\n客户经理反馈意见（请据此调整报告）：\n${feedback}`;
  }
  return callLLM(skill, userPrompt);
}

async function runFullAnalysis(holdingsText, tradesText, customerName = "客户", feedback = null) {
  const assetResult = await analyzeAssets(holdingsText, feedback);
  const tradeResult = await analyzeTrades(tradesText, feedback);
  const summary = await generateSummary(assetResult, tradeResult, customerName, feedback);

  return {
    asset_analysis: assetResult,
    trade_analysis: tradeResult,
    summary,
  };
}

module.exports = { runFullAnalysis };
